viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -526,31 +526,40 @@ void SLAMServiceImpl::RunSLAM() {
 }
 
 std::string SLAMServiceImpl::GetNextDataFileOffline() {
-    if (!b_continue_session) {
-        return "";
-    }
-    if (file_list_offline.size() == 0) {
-        file_list_offline = viam::io::ListSortedFilesInDirectory(path_to_data);
-    }
-    // We're setting the minimum required files to be two for the following
-    // reasons:
-    // 1. Cartographer needs at least two PCD files to work properly.
-    // 2. A .DS_Store file is frequently added to the data directory when
-    // a user opens the directory on osx.
-    // Expecting a minimum of 3 files solves both problems without having to
-    // loop over and count the number of actual data files in the data
-    // directory.
-    if (file_list_offline.size() <= 2) {
-        throw std::runtime_error("not enough data in data directory");
-    }
-    if (current_file_offline == file_list_offline.size()) {
+    while (b_continue_session) {
+        if (file_list_offline.size() - current_file_offline <= 2) {
+            file_list_offline =
+                viam::io::ListSortedFilesInDirectory(path_to_data);
+        }
+        // We're setting the minimum required files to be two for the following
+        // reasons:
+        // 1. Cartographer needs at least two PCD files to work properly.
+        // 2. A .DS_Store file is frequently added to the data directory when
+        // a user opens the directory on osx.
+        // Expecting a minimum of 3 files solves both problems without having to
+        // loop over and count the number of actual data files in the data
+        // directory.
+        if (file_list_offline.size() > 2) {
+            // If the current file is the last in the file list, skip it, since
+            // it's possible it's still being written to.
+            if (current_file_offline != file_list_offline.size() - 1) {
+                const auto to_return = file_list_offline[current_file_offline];
+                if (to_return != "") {
+                    current_file_offline++;
+                    return to_return;
+                }
+            }
+        }
+
+        // Sleep for a short period of time if we didn't find any files. This is
+        // to prevent cartographer from using up too much CPU in this loop when
+        // there's no new data coming in.
+        std::this_thread::sleep_for(data_rate_ms);
```

Comment:
This is not strictly necessary but it seemed nice. We could add this to the data process loop too in `viam-cartographer.go`.

Reply:
Will data_rate_ms be zero for actual offline mode? If not, this will cause actual offline to run unnecessarily slowly.

Also, I don't understand why we would add this to the data process loop in viam-cartographer.go, since that loop doesn't have to wait for new data to come in.

Reply:
It shouldn't cause this to run slow because we only sleep if we were unable to find more files. For offline mode, this should only happen when we reach the end of the dataset. For the replay camera, this could happen if there replay camera is unable to keep up with cartographer and files are not being written fast enough. I added this because after this change, we never stop SLAM in offline mode/replay camera mode, so this should make this loop less resource-intensive once we've reached the end of the dataset.

Reply:
Ah, got it. I think it's a little confusing to use data_rate_ms here, since how long we should wait doesn't have to do with the data rate. But also we'll stop writing to files in full modularization, so we don't need to think too carefully about what's the right thing to do here.